src/main/java/io/appium/java_client/remote/AppiumCommandExecutor.java

```diff
@@ -166,13 +211,43 @@ private Response createSession(Command command) throws IOException {
         setCommandCodec(new AppiumW3CHttpCommandCodec());
         refreshAdditionalCommands();
         setResponseCodec(dialect.getResponseCodec());
-        return result.createResponse();
+        Response response = result.createResponse();
+
+        setDirectConnect(response);
+
+        return response;
     }
 
     public void refreshAdditionalCommands() {
         getAdditionalCommands().forEach(this::defineCommand);
     }
 
+    @SuppressWarnings("unchecked")
+    private void setDirectConnect(Response response) throws MalformedURLException {
```

Comment:
we must wrap this exception and provide a proper error message to the user if the server did not return correct values to construct the url